### ISSUE TITLE
Since frontier management in storage

### ISSUE THREAD

**Issue by petrosagg** [MaterializeInc/materialize#15687]
There have been a lot of discussions lately about the bugs we've hit in storage that are related in one way or another with how we manage the since frontier of collections and shards and who triggers compaction for them. This issue is an attempt to clearly explain what exactly caused the issues and what a solution could look like.

### Small COMPUTE digression

Before explaining how the problem in storage happened it's helpful to have some context on how the exact same problem is being handled successfully in compute. The dual of a storage collection in compute is an in-memory index. The two types of objects are similar in that they are both differential traces of a collection and have an associated since and upper frontier that define what is able to be read and written.

Indexes are objects that live in `computed` processes and are created when the compute controller sends a `CreateDataflows` command to `computed` instances with an index description in it. When a `computed` cluster receives such command it renders the requested dataflow and starts producing data for the index, advancing the upper frontier as it goes.

After an index has been installed, more dataflows can use this index on the same cluster. Each of those dependent dataflows might make progress at different rates and therefore read data at different rates. Additionally, external read constraints might be placed along the since frontier of an index. This where we arrive at the correctness property that must be maintained:

**The since frontier must not advance past the requirements of any of the readers**

But what are those requirements and who knows them? The answer is: the compute controller. The compute controller has the authoritative and global view of all the dependencies of an index, including arbitrary read policies set by the user ,and can make an informed decision about what the since frontier should be and record that decision with an `AllowCompaction` command. Here we arrive at the mechanism that ensures the correctness property:

**No index is allowed to compact UNLESS it has received an `AllowCompaction` command**

### The storage/persist correctness issue

The system described above sounds pretty good and we should be able to apply it as-is in storage. In fact, the storage controller continuously calculates and sends `AllowCompaction` commands to `storaged` processes for them to compact their shards at. The frontier included in those commands is guaranteed to take into account the needs of all the readers.

How is it possible then that we hit since frontier problems if we have this nice mechanism that provides a global bound on what the since frontier should be?

There are two pieces to the answer. First, `storaged` instances completely ignore `AllowCompaction` commands and only react to them when they reach the empty frontier to clean up resources. Secondly, `persist` does not allow you to read a shard without also downgrading its since. This means that instead of having every single `persist` handle follow religiously the `AllowCompaction` commands that the storage controller prepared for them, we aggressively downgrade as much as we can just as a mere side effect of reading the data and rely on the storage controller to hold onto additional read handles that must not fail to heartbeat in order to cancel these effects out.

### Why can't storage be like compute?

The reader might now wonder, if the `AllowCompaction` commands are already wired up why don't we just delete all the code that downgrades since frontiers arbitrarily and only downgrade them when the storage controller says so? There isn't a good reason that isn't the case except for the fact that it's impossible to read data without downgrades using the current `persist` API. The good news is that the change needed to make this happen are minimal, [as showcased by this PR](https://github.com/MaterializeInc/materialize/pull/15683). If we do that not only we fix the correctness issues and stop relying on clocks, but we also make the compute and storage look more similar, increasing the probability for code reuse in the future.

### Ok, but what about the critical/non-critical handle proposal?

For that I will refer back to the index/shard duality and claim that just like indexes don't need any sort of special critical/non-critical handle to manage their since frontiers we shouldn't need anything like that in storage. The only reason we've seen bugs with storage shards but not (as many) with compute indexes is because we ignored the solution that was there all along and it's a single simple rule:

> Only downgrade when receiving an `AllowCompaction` command

In other words, the only reason we relied on read leases for correctness is because we did it to ourselves.

Therefore my proposal is to merge the PR linked in the section above so that we can decouple reading data from a shard from its since frontier and start honoring the `AllowCompaction` commands that are now dropped on the floor. This will allow persist to keep its simple API and not have two separate classes of read handles.

**Comment by petrosagg** [comment]
Turns out that after I wrote all this down I found a hole in my proposal whose solution is.. critical/non-critical handles, so I think I changed my mind!

The problem with the plan I outlined above, summarized as "keep handles as-is, call `downgrade_since` in response to `AllowCompaction` commands", is that the `AllowCompaction` commands are only sent to the `storaged` processes that are producing the data. However, there are more handles that live in the compute instances (e.g as part of a persist source) and those never see the storage related `AllowCompaction` commands, meaning that they would hold back compaction as long as they were alive.

The only way to rectify that would be to broadcast `AllowCompaction` commands to any process that has a handle but that sounds arbitrary, so the only solution left is that those handles should be non-critical

**Comment by lluki** [comment]
> However, there are more handles that live in the compute instances (e.g as part of a persist source) and those never see the storage related AllowCompaction commands, meaning that they would hold back compaction as long as they were alive.

I think adapting the compute model would mean:
 * Readers downgrade as soon as they can
 * Downgrades from readers are silent: They don't cause a message to be sent to envd or anywhere else
 * The envd controller determines using FrontierUppers of dataflows that read from these collections when it is safe to downgrade the per-collection since
 * If this per collection since has advanced, sends AllowCompaction.
 * Computeds are then free to compact

Afaik this should also work with storage collections, the only difference I see is that any process could potentially do storage compaction, while with arrangements it's very clear that the process that has it in memory has to do the compaction. Imho the decision where and when to do compaction is a policy decision best made with global information available, which would be in the StorageController.

I do like this model in general, the one part I don't like is that we use FrontierUppers to determine the per collection since. If a dataflow uses a lot of time, even though it's not reading from the collection anymore, we don't notice until the FrontierUpper is sent. I'd be in favor of an extra "read since downgrade" message. 

@teskje probably has opinions on this as well.

PS: I asked some questions some related Qs on how it works (soon'ish) in compute [here](https://github.com/MaterializeInc/materialize/pull/15614#issuecomment-1290250653).

**Comment by petrosagg** [comment]
> Readers downgrade as soon as they can

This is what we do now but unlike indexes where the resource being downgraded is local to the process performing the downgrades with collections we rely on timeouts to ensure that a crashed process doesn't permanently hold back the since frontier.

I thought about this a bit more since yesterday and I think we can remove the concept of a read hold. Why does the mere existence of a read handle need to have a say on what the since should be? If no handle had any hold and any handle could downgrade since at any time then everything works out with the existing controller model, reacting to `AllowCompaction` commands as appropriate.

> Imho the decision where and when to do compaction is a policy decision best made with global information available, which would be in the StorageController.

Yes, I think there is general agreement on this one

> If a dataflow uses a lot of time, even though it's not reading from the collection anymore, we don't notice until the FrontierUpper is sent. I'd be in favor of an extra "read since downgrade" message.

Hm, is this possible to do this? If a dataflow has finished reading some time `t` from an input and then takes a lot of time to produce the output is it correct to downgrade the since before the output for `t` is out? What happens we compact while the computation is in progress but then `computed` crashes?

**Comment by teskje** [comment]
Not sure how relevant that is to finding a solution to storage, but I would like to add one detail of how compaction works with compute collections. It is true that the compute controller uses `AllowCompaction` commands to hold back compaction of indexes according to the needs of external readers (i.e. readers outside compute). But it does *not* hold back compaction for readers inside compute, because it doesn't need to. What I mean by this is that the following can happen:
* There are two replicas in a cluster, both maintaining an index _I_ and a dataflow _D_ reading from that index.
* One replica is faster than the other. It reports to the compute controller that the write frontier of _D_ has advanced.
* In response, the compute controller advances the read frontier of _D_ as well (based on _D_'s read policy).
* With the read frontier of _D_ advanced, the read frontier of its dependency _I_ can also advance.
* The compute controller sends an `AllowCompaction` command to all replicas, containing the new read frontier of _I_.
* So the slow replica receives a command that allows it to compact _I_ to its new read frontier, even though it might not have read from _I_ that far.

The crucial detail here is that within timely, readers of arrangements hold back compaction (same as persist `ReadHandle`s currently do for persist shards). So even if the compute controller allows compaction to some time, a replica's arrangements are only compacted as far as its internal readers allow it.

Maybe everyone already understood this, in which case sorry for the noise!

> What happens we compact while the computation is in progress but then computed crashes?

Yes, because of this we cannot allow dependencies' sinces to downgrade beyond the sinces of their dependants. If `computed` crashes we need to be able to rehydrate dataflows outputs as-of their last since, which requires the dataflow inputs to also still be readable at that since.

What we can do is relax per-replica read holds on dataflow inputs as soon as the replicas have finished reading from those inputs at some time. For this, we'd need a new kind of compute response that actually transmits that information. But we would still need to rely on `FrontierUppers` to determine the "implied capability" of the collection.

**Comment by petrosagg** [comment]
> The crucial detail here is that within timely, readers of arrangements hold back compaction (same as persist ReadHandles currently do for persist shards). So even if the compute controller allows compaction to some time, a replica's arrangements are only compacted as far as its internal readers allow it.

What a great point! I hadn't thought about what happens with multiple replicas but this makes a ton of sense. We don't have (yet) replicated storage ingestion but there is no reason we shouldn't have in the future.

To me this sounds like the final point in favor of the  critical/non-critical handles design because an active lagged replica will need to hold back the since while it catches up.

**Comment by petrosagg** [comment]
Thank you @lluki and @teskje for the input, this are so much clearer in my head now! I think I'll close this issue since critical/non-critical handles combined with honoring `AllowCompaction` commands for the critical one solve the problem and also open the possibility for replicated ingestion in the future
